eli5/keras.py

```diff
@@ -0,0 +1,465 @@
+    """
+    Explain the prediction of a Keras image classifier.
+
+    See :func:`eli5.explain_prediction` for more information about the ``estimator``,
+    ``doc``, ``target_names``, and ``targets`` parameters.
+
+    Parameters
+    ----------
+    estimator : keras.models.Model
+        Instance of a Keras neural network model.
+
+    doc : numpy.ndarray
+        An input image as a tensor to ``estimator``, for example a ``numpy.ndarray``.
+
+        The tensor must be of suitable shape for the ``estimator``. 
+        For example, some models require input images to be 
+        rank 4 in format `(batch_size, dims, ..., channels)` (channels last)
+        or `(batch_size, channels, dims, ...)` (channels first), 
+        where batch size is 1 for a single image.
+
+    target_names : list, optional
+        *Not Implemented*. 
+
+        Names for classes in the final output layer.
+
+    targets : list[int], optional
+        Prediction ID's to focus on.
+
+        *Currently only the first prediction from the list is explained*. 
+        The list must be length one.
+
+        If None, the model is fed the input image and its top prediction 
+        is taken as the target automatically.
+
+    layer : int or str or keras.layers.Layer, optional
+        The activation layer in the model to perform Grad-CAM on,
+        a valid keras layer name, layer index, or an instance of a Keras layer.
+        
+        If None, a suitable layer is attempted to be retrieved. 
+        See :func:`eli5.keras.search_layer_backwards` for details.
+
+    Returns
+    -------
+    expl : Explanation
+        A :class:`eli5.base.Explanation` object 
+        with the ``image`` and ``heatmap`` attributes set.
+    """
+    # TODO: implement target_names
+    # FIXME: Could doc be a Tensorflow object, not just a numpy array?
+    validate_doc(estimator, doc)
+    activation_layer = get_activation_layer(estimator, layer)
+    
+    heatmap, predicted_idx, score = grad_cam(estimator, doc, targets, activation_layer)
+    # TODO: consider renaming 'heatmap' to 'visualization'/'activations'
+    # (the output is not yet a heat map)
+
+    print('Predicted class: %d' % predicted_idx)
+    print('With probability: %f' % score)
+
+    # TODO: consider passing multiple images in doc to perform grad-cam on multiple images
+    doc = doc[0] # rank 4 batch -> rank 3 single image
+    image = keras.preprocessing.image.array_to_img(doc) # -> PIL image
+
+    return Explanation(
+        estimator.name, # might want to replace this with something else, eg: estimator.summary()
+        description=DESCRIPTION_KERAS,
+        error='',
+        method='Grad-CAM',
+        is_regression=False, # TODO: classification vs regression model
+        highlight_spaces=None, # might be relevant later when explaining text models
+        image=image, # PIL image
+        heatmap=heatmap, # 2D [0, 1] numpy array
+    )
+
+
+def validate_doc(estimator, doc):
```

Comment:
Same question regarding visibility in API docs - if these functions are visible, let's hide them.

Reply:
Perhaps `grad_cam()` and `grad_cam_backend()` functions should be kept public? This way the user can check the values of the gradients, weights, activations, etc.

Reply:
What do you think about making eli5.keras a package, and then have eli5.keras.grad_cam module, + maybe eli5.keras.utils?

Reply:
if grad_cam is not keras-specific, then it can be eli5.gradcam as well.

Reply:
Sounds like a good way to partition things.

Though I am not sure whether I should have `eli5.keras.gradcam` or `eli5.gradcam`. I could have both, placing code that could be re-used in `eli5.gradcam` (see @lopuhin's comment about pytorch support just down below), and Keras-specific code in the other.

Reply:
I think it's fine if we eventually have both ``eli5.gradcam`` and ``eli5.keras.gradcam``, but for now I think it's fine to just have `eli5.keras.gradcam`` and factor out common code when we add pytorch support.

Reply:
> Perhaps grad_cam() and grad_cam_backend() functions should be kept public? This way the user can check the values of the gradients, weights, activations, etc.

Sure, if you think they can be useful let's keep them public.